cpp/src/arrow/builder.cc

```diff
@@ -920,6 +957,13 @@ Status DictionaryBuilder<T>::AppendArray(const Array& array) {
   return Status::OK();
 }
 
+Status DictionaryBuilder<NullType>::AppendArray(const Array& array) {
+  for (int64_t i = 0; i < array.length(); i++) {
+    RETURN_NOT_OK(AppendNull());
```

Comment:
Per my comment above, when we handle null as a category value, this will need to append a 0 here